### ISSUE TITLE
IoULoss and GIoULoss fail with retinanet

### ISSUE THREAD

**Issue by melikovk** [open-mmlab/mmdetection#2417]
The following command fails with a tensor size mismatch RuntimeError if SmoothL1 loss is replaced with IoULoss or GIoULoss, while BoundedIoULoss works fine
`python tools/train.py configs/retinanet_r50_fpn_1x_py`
I have attached the full error log for information.
The error seems to stem from the fact that iou_loss and gios_loss output tensor with the shape [Npos] while bbox_weights for anchor_heads have the shape [Npos, 4]. It seems that shape of bbox_weights can be changed to match. I would gladly try to fix the issue. It is not clear to me though if you still work on v1.1 or you are focused on rewriting version 2

Thank you
[error_iou.log](https://github.com/open-mmlab/mmdetection/files/4451073/error_iou.log)

**Comment by hellock** [comment]
SmoothL1Loss and IoULoss require different inputs, we proposed a unified API in #2156, but it needs improvement. It is planned as a feature of v2.0.

**Comment by melikovk** [comment]
Let me know if I can help with it. I will be glad to contribute in any way.

**Comment by iamRyanChia** [comment]
When will the 2.0 version be released? Can you give a rough date? @ @hellock

**Comment by iamRyanChia** [comment]
> SmoothL1Loss and IoULoss require different inputs, we proposed a unified API in #2156, but it needs improvement. It is planned as a feature of v2.0.

When will the 2.0 version be released? Can you give a rough date? @hellock

**Comment by melikovk** [comment]
I have tried to checkout v2.0 and run train.py but got an ImportError poteintialy due to circular import. I was wandering if this is to be expected at this stage. I also looked into the code for box regression loss calculation and for what it is worth I  think if you will make log transform of the bbox sizes in bbox2delta optional it will simplify use of alternative loss functions such as IOU. Since IOU is scale invariant you will be able to pass anchor normalized predictions and targets to the IOU loss. I am also unclear why you need target means in bbox2delta, since those are simply constant offsets and should be accommodated by the last layer bias. It also seems to me that  target stds act as relative weights for box offset and size contribution and are probably better moved to the loss function (they probably make more sense for smothL1 and L2 loss than IOU losses). Sorry for the long comment and hopefully this is more helpful than distracting. If needed I can turn this into a PR.

**Comment by xvjiarui** [comment]
Hi @melikovk 
Sorry for the late update. After merged #2422 #2429 #2480 , we are working on #2156. It would be merged very soon. IoULoss and GIoULoss would be supported by then. 
Target means/stds are used to normalized regression targets, as the practice in [R-CNN](https://arxiv.org/abs/1311.2524). 
Your help is truly appreciated. You are very welcome to review these PRs.

**Comment by melikovk** [comment]
Hi @xvjiarui
Thank you for the update. No need for apologies, I perfectly understand that you have other things to do. Could you, please, clarify if the error I am getting with train.py in v2.0 is expected or I should be able to run training using v2.0 branch? Thank you very much.

**Comment by hellock** [comment]
> I have tried to checkout v2.0 and run train.py but got an ImportError poteintialy due to circular import. I was wandering if this is to be expected at this stage. I also looked into the code for box regression loss calculation and for what it is worth I think if you will make log transform of the bbox sizes in bbox2delta optional it will simplify use of alternative loss functions such as IOU. Since IOU is scale invariant you will be able to pass anchor normalized predictions and targets to the IOU loss. I am also unclear why you need target means in bbox2delta, since those are simply constant offsets and should be accommodated by the last layer bias. It also seems to me that target stds act as relative weights for box offset and size contribution and are probably better moved to the loss function (they probably make more sense for smothL1 and L2 loss than IOU losses). Sorry for the long comment and hopefully this is more helpful than distracting. If needed I can turn this into a PR.

Target means and stds are used for regression losses like L1 and SmoothL1, which has been a typical convention. For IoU losses, these two parameters are unnecessary but we kept it to make the api more unified and clean. Your suggestion to make the log transform optional is great, actually we propose a more flexible solution in #2480, which adds a bbox coder class to perform the encoding and decoding. Users can define any transforms (like in IoU losses or anchor free detectors).

**Comment by melikovk** [comment]
Thank you. The solution in #2480 is indeed a great flexible solution. I am not sure I like how it is used in #2156. I think that adding ability to remove log transform would be more transparent and as a bonus would not require you to back-propagate the inverse transform. But in the end this is a matter of preference as this probably will have minor time and memory impact. Thank you for the great work you are doing.
